bulk/repo-access-api.adoc

```diff
@@ -1,155 +1,335 @@
-{
-  "versionHash": "af945d129c476",
-  "baseVersionHash": "39a15fd49c867",
-  "nodes": [
-    // only nodes that changed between the two versions
-  ]
-}
-----
 
-Also optional.
-If the server only supports bulk model read/write operations then it returns 404 and the client will fall back to the bulk endpoints.
+[[store, store]]
+=== store: Put nodes into repository
+Creates new nodes, or updates existing nodes in the repository.
+
+We always process one node in its entirety, i.e. we cannot update parts of the node with this API.
+
+A node id referenced as parent, containment, reference, or annotation _can_ be mentioned in the same request, but _can_ be omitted if a node with that id already exists in the repository.
+This way, we can move subtrees and add arbitrary references without sending unchanged nodes.
+
+We consider a node to be _updated_ if a node with the same id already exists in the repository.
+Otherwise, we consider that node to be _new_.
+
+==== Parameters
+[[store.nodes]]
+`nodes`:: <<{serialization}.adoc#SerializationChunk, SerializationChunk>> containing all nodes to store to the repository.
+
+[[store.mode]]
+`mode`:: Kind of storage accepted by the repository.
+Optional parameter, defaults to `replace`.
++
+#TODO: Discuss names#
+
+[[store.mode.replace, `replace`]]
+`replace`::: Accept both new and updated nodes.
+[[store.mode.create]]
+`create`::: Only accept new nodes.
+The call fails without any changes to the repository if any node id from `nodes` already exists in the repository.
+The call accepts already existing node ids only referenced as child, reference, or annotation.
+[[store.mode.update]]
+`update`::: Only accept existing nodes.
+The call fails without any changes to the repository if any node id from `nodes` does not exist in the repository.
+#TODO: Do we want this mode?#
+
+==== Semantics
+After completing this call, all sent nodes MUST have exactly the sent contents (except `parent`, see below #TODO correct?#) in the repository.
+We must send containments/annotations from the parent's view, because we need to know the containment and index of the contained node/annotation within its parent.
+
+If we move a contained/annotation node `C` from its previous parent `A` to its new parent `B`, we MUST send `B`, and MAY omit `A`.
+This means we can have implicit changes in `A`.
+
+The whole call fails, without any changes to the repository, if it would lead to a malformed model (i.e. cycles in containment hierarchy, same node contained in different parents, parent and corresponding containment/annotation don't match #TODO: Add reference to model validation#).
+
+The repository SHOULD NOT #TODO: or MUST NOT?# validate the "meta-semantics" of the sent nodes, e.g. whether a Concept instance without parent is a partition, a feature instance can be part of its node instance, or a node mentioned as annotation is an Annotation instance.
+Rationale: We want to support storing invalid models, so the user does not lose data.
```

Comment:
Jos:: The repository should work without metamodel knmowledge, but some kind of correctness will be needed,
otherwise many clients won't work.
We should probably cross-reference with the validation document to see which rules are to be maintained by the repository.
Jos:: If the metamodel is know n, the repository can check many extra things.
That _could_ be beneficiary for all clients.
Maybe we should have optional validations on the server?